File: src/main/java/club/sk1er/patcher/tweaker/asm/BakedQuadTransformer.java
/*
 * Copyright © 2020 by Sk1er LLC
 *
 * All rights reserved.
 *
 * Sk1er LLC
 * 444 S Fulton Ave
 * Mount Vernon, NY
 * sk1er.club
 */

package club.sk1er.patcher.tweaker.asm;

import club.sk1er.patcher.tweaker.transform.PatcherTransformer;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.tree.ClassNode;
import org.objectweb.asm.tree.InsnList;
import org.objectweb.asm.tree.InsnNode;
import org.objectweb.asm.tree.MethodInsnNode;
import org.objectweb.asm.tree.MethodNode;
import org.objectweb.asm.tree.VarInsnNode;

public class BakedQuadTransformer implements PatcherTransformer {
    /**
     * The class name that's being transformed
     *
     * @return the class name
     */
    @Override
    public String[] getClassName() {
        return new String[]{"net.minecraft.client.renderer.block.model.BakedQuad"};
    }

    /**
     * Perform any asm in order to transform code
     *
     * @param classNode the transformed class node
     * @param name      the transformed class name
     */
    @Override
    public void transform(ClassNode classNode, String name) {
        for (MethodNode methodNode : classNode.methods) {
            if (methodNode.name.equals("equals")) {
                System.out.println("User has Frames+, not creating BakedQuad#equals.");
                return;
            }
        }

        System.out.println("Creating BakedQuad#equals.");
        MethodNode equals = new MethodNode(Opcodes.ACC_PUBLIC, "equals", "(Ljava/lang/Object;)Z", null, null);
        equals.instructions.add(getEquals());
        classNode.methods.add(equals);
    }

    private InsnList getEquals() {
        InsnList list = new InsnList();
        list.add(new VarInsnNode(Opcodes.ALOAD, 0));
        list.add(new VarInsnNode(Opcodes.ALOAD, 1));
        list.add(new MethodInsnNode(Opcodes.INVOKESTATIC,
            "club/sk1er/patcher/util/world/block/BlockUtil",
            "bakedQuadEquals",
            "(Lnet/minecraft/client/renderer/block/model/BakedQuad;Ljava/lang/Object;)Z",
            false));
        list.add(new InsnNode(Opcodes.IRETURN));
        return list;
    }
}

File: src/main/kotlin/club/sk1er/patcher/asm/render/block/BakedQuadTransformer.kt
/*
 * Copyright © 2020 by Sk1er LLC
 *
 * All rights reserved.
 *
 * Sk1er LLC
 * 444 S Fulton Ave
 * Mount Vernon, NY
 * sk1er.club
 */

package club.sk1er.patcher.asm.render.block

import club.sk1er.hookinjection.getInstructions
import club.sk1er.patcher.tweaker.transform.PatcherTransformer
import club.sk1er.patcher.util.world.render.block.BlockUtil
import org.objectweb.asm.Opcodes
import org.objectweb.asm.tree.ClassNode
import org.objectweb.asm.tree.MethodNode

class BakedQuadTransformer : PatcherTransformer {
    override fun getClassName() = arrayOf("net.minecraft.client.renderer.block.model.BakedQuad")

    override fun transform(classNode: ClassNode, name: String) {
        val equals = MethodNode(Opcodes.ACC_PUBLIC, "equals", "(Ljava/lang/Object;)Z", null, null)
        equals.instructions.add(getEquals(equals))
        classNode.methods.add(equals)
    }

    private fun getEquals(method: MethodNode) = getInstructions {
        of(BlockUtil::bakedQuadEquals)
        into(method)
        params(0, 1)
        keepReturns
    }
}